File: Combinations.java
import java.util.ArrayList;
import java.util.List;

/**
 * Created by yhf on 15/5/23.
 */

/*
Given two integers n and k, return all possible combinations of k numbers out of 1 ... n.

For example,
If n = 4 and k = 2, a solution is:

[
  [2,4],
  [3,4],
  [2,3],
  [1,2],
  [1,3],
  [1,4],
]
*/

public class Combinations {
    public static List<List<Integer>> combine(int n, int k) {
        List<Integer> cur = new ArrayList<>();
        List<List<Integer>> res = new ArrayList<>();
        if (n <= 0) {
            return res;
        }
        combine(n, k, 1, cur, res);
        return res;
    }

    private static void combine(int n, int k, int idx, List<Integer> cur, List<List<Integer>> res) {
        if (cur.size() == k) {
            res.add(new ArrayList<>(cur));
            return;
        }

        if (idx > n) {
            return;
        }

        combine(n, k, idx + 1, cur, res);
        cur.add(idx);
        combine(n, k, idx + 1, cur, res);
        cur.remove(cur.size() - 1);
    }

    public static void main(String[] args) {
        List<List<Integer>> res = combine(4, 2);
    }
}

File: Combinations.py
"""
Given two integers n and k, return all possible combinations of k numbers out of 1 ... n.

For example,
If n = 4 and k = 2, a solution is:

[
  [2,4],
  [3,4],
  [2,3],
  [1,2],
  [1,3],
  [1,4],
]
"""


class Solution:
    # @param {integer} n
    # @param {integer} k
    # @return {integer[][]}
    def combine(self, n, k):
        self.__res = []
        if n <= 0 or k <= 0:
            return self.__res

        def __combine(cur, idx):
            if len(cur) == k:
                self.__res.append(list(cur))
                return
            if idx > n:
                return
            cur.append(idx)
            __combine(cur, idx + 1)
            cur.pop()
            __combine(cur, idx + 1)

        __combine([], 1)
        return self.__res


if __name__ == '__main__':
    print Solution().combine(4, 2)
